examples/light-switch-app/silabs/src/LightSwitchMgr.cpp

```diff
@@ -26,18 +26,123 @@
 
 #include "AppConfig.h"
 #include "AppEvent.h"
+#include "AppTask.h"
 
 #include <app-common/zap-generated/attributes/Accessors.h>
 #include <app-common/zap-generated/cluster-objects.h>
 #include <app/clusters/switch-server/switch-server.h>
 #include <platform/CHIPDeviceLayer.h>
+#include <cmsis_os2.h>
+#include <platform/silabs/platformAbstraction/SilabsPlatform.h>
 
 using namespace chip;
 using namespace chip::app;
 using namespace chip::app::Clusters;
+using namespace chip::DeviceLayer;
+using namespace chip::DeviceLayer::Silabs;
 
 LightSwitchMgr LightSwitchMgr::sSwitch;
 
+AppEvent CreateNewEvent(AppEvent::AppEventTypes type)
+{
+    AppEvent aEvent;
+    aEvent.Type                      = type;
+    aEvent.Handler                   = LightSwitchMgr::GeneralEventHandler;
+    LightSwitchMgr * lightSwitch         = &LightSwitchMgr::GetInstance();
+    aEvent.LightSwitchEvent.Context  = lightSwitch;
+    return aEvent;
+}
+
+void LightSwitchMgr::Timer::Start()
+{
+    // Starts or restarts the function timer
+    if (osTimerStart(mHandler, pdMS_TO_TICKS(100)) != osOK)
+    {
+        SILABS_LOG("Timer start() failed");
+        appError(CHIP_ERROR_INTERNAL);
+    }
+
+    mIsActive = true;
+}
+
+void LightSwitchMgr::Timer::Timeout()
+{
+    mIsActive = false;
+    if (mCallback)
+    {
+        mCallback(*this);
+    }
+}
+
+void LightSwitchMgr::HandleLongPress()
+{
+    AppEvent event;
+    event.Handler             = GeneralEventHandler;
+    LightSwitchMgr * lightSwitch  = &LightSwitchMgr::GetInstance();
+    event.LightSwitchEvent.Context  = lightSwitch;
+    if (mDownPressed)
+    {
+        if (!mResetWarning)
+        {
+            // Long press button up: Reset warning!
+            event.Type = AppEvent::kEventType_ResetWarning;
+            AppTask::GetAppTask().PostEvent(&event);
+        }
+    }
+}
+
+void LightSwitchMgr::OnLongPressTimeout(LightSwitchMgr::Timer & timer)
+{
+    LightSwitchMgr * app = static_cast<LightSwitchMgr *>(timer.mContext);
+    if (app)
+    {
+        app->HandleLongPress();
+    }
+}
+
+LightSwitchMgr::Timer::Timer(uint32_t timeoutInMs, Callback callback, void * context) : mCallback(callback), mContext(context)
```

Comment:
timeoutInMs is not used in the configuration - i think the timer logic might need to be updated.

Reply:
Can we create a follow up ticket to finish cleaning this up?